lib/charms/vault_k8s/v0/vault_kv.py

```diff
@@ -0,0 +1,682 @@
+        secret = self.model.get_secret(id=unit_credentials)
+        secret_content = secret.get_content(refresh=True)
+        role_id = secret_content["role-id"]
+        role_secret_id = secret_content["role-secret-id"]
+
+        self._configure(vault_url, ca_certificate, mount, role_id, role_secret_id)
+
+        self.unit.status = ActiveStatus()
+
+    def _on_gone_away(self, event: vault_kv.VaultKvGoneAwayEvent):
+        self.unit.status = BlockedStatus("Waiting for vault-kv relation")
+
+    def _configure(
+        self,
+        vault_url: str,
+        ca_certificate: str,
+        mount: str,
+        role_id: str,
+        role_secret_id: str,
+    ):
+        pass
+
+    def _on_update_status(self, event):
+        # Check somewhere that egress subnet has not changed i.e. pod has not been rescheduled
+        # Update status might not be the best place
+        binding = self.model.get_binding("vault-kv")
+        if binding is not None:
+            egress_subnets = [str(subnet) for subnet in self.model.get_binding(relation).network.egress_subnets][0].subnet]
+            egress_subnets.append(str(self.model.get_binding(relation).network.interfaces[0].subnet))
+            relation = self.model.get_relation(relation_name="vault-kv")
+            self.interface.request_credentials(relation, egress_subnets, self.get_nonce())
+
+    def get_nonce(self):
+        secret = self.model.get_secret(label=NONCE_SECRET_LABEL)
+        nonce = secret.get_content(refresh=True)["nonce"]
+        return nonce
+
+
+if __name__ == "__main__":
+    main(ExampleRequirerCharm)
+```
+
+You can integrate both charms by running:
+
+```bash
+juju integrate <vault provider charm> <vault requirer charm>
+```
+"""
+
+import json
+import logging
+from collections.abc import Iterable, Mapping
+from dataclasses import dataclass
+from typing import Any, Dict, List
+
+import ops
+from interface_tester.schema_base import DataBagSchema  # type: ignore[import-untyped]
+from pydantic import BaseModel, Field, Json, ValidationError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "591d6d2fb6a54853b4bb53ef16ef603a"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 11
+
+PYDEPS = ["pydantic", "pytest-interface-tester"]
+
+
+class LogAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend a prefix to all log lines."""
+
+    prefix = "vault_kv"
+
+    def process(self, msg, kwargs):
+        """Decides the format for the prepended text."""
+        return f"[{self.prefix}] {msg}", kwargs
+
+
+logger = LogAdapter(logging.getLogger(__name__), {})
+
+
+class VaultKvProviderSchema(BaseModel):
+    """Provider side of the vault-kv interface."""
+
+    vault_url: str = Field(description="The URL of the Vault server to connect to.")
```

Comment:
Isn't there a pydantic type for this?

Reply:
I created a tickets to use pydantic validation